src/Common/PoolBase.h

```diff
@@ -184,6 +211,8 @@ class PoolBase : private boost::noncopyable
     /** The maximum size of the pool. */
     unsigned max_items;
 
+    BehaviourOnLimit behaviour_on_limit;
```

Comment:
How about different model here?
Soft and hard limit.
If pool is under soft limit, it allocates more. When some connection if freed, pool closes excess connections.
If pool is close to hard limit. It waits no more than connection timeout when limit hits hard level.

Reply:
what advantages it would have? from perf standpoint any waiting is bad I think.

Reply:
That with those 2 values you is able configure system in a two ways: when it waits a vacant connection and when it doesn't wait any. And the most important part taht you will never go into infinite connections count.

>  from perf standpoint any waiting is bad I think.
What do you have as alternative way?

Ofcourse you could fail request instantly. But it is unlogical, since there is a settings connection_timeout. 

Imagine that there are a lot of requests to the S3. If you just fail all the requests above the limit the cluster would stop unable to do a progress, that would be denial of service problem. If you slow down the progress by waiting the connections the progress will continue, that would be performance issue with some ways how to mitigate it by changing load pattern from user side or by changing settings from our side. Sounds like gracefull degradation.